### PR TITLE
Pre Depth pass

### PR DESCRIPTION
**Pull request by fjz345**
Before forward pass there is a pre depth pass, forward pass does not write to any depth now.

This was tested by me with Pix.

### CODE REVIEW

File: StortSpelProjekt/Engine/src/Renderer/DX12Tasks/PreDepthRenderTask.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
+
+#include "stdafx.h"
+#include "PreDepthRenderTask.h"
+
+#include "../Mesh.h"
+#include "../Transform.h"
+#include "../RootSignature.h"
+#include "../CommandInterface.h"
+#include "../SwapChain.h"
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/DX12Tasks/PreDepthRenderTask.cpp:9]: Do you really need the swapchain here? Check what includes you actually use

File: StortSpelProjekt/Engine/src/Renderer/DX12Tasks/PreDepthRenderTask.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+	const DirectX::XMMATRIX* viewProjMatTrans = m_pCamera->GetViewProjectionTranposed();
+
+	
+	/*
+	commandList->ResourceBarrier(1, &CD3DX12_RESOURCE_BARRIER::Transition(
+		m_pSwapChain->GetRenderTarget(m_BackBufferIndex)->GetResource()->GetID3D12Resource1(),
+		D3D12_RESOURCE_STATE_PRESENT,
+		D3D12_RESOURCE_STATE_DEPTH_WRITE));
+		*/
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/DX12Tasks/PreDepthRenderTask.cpp]: Remove, or do you need this?

File: StortSpelProjekt/Engine/src/Renderer/DX12Tasks/PreDepthRenderTask.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		// Draws all entities with ModelComponent + TransformComponent
+		drawRenderComponent(mc, tc, viewProjMatTrans, commandList);
+	}
+
+	/*
+	commandList->ResourceBarrier(1, &CD3DX12_RESOURCE_BARRIER::Transition(
+		m_pSwapChain->GetRenderTarget(m_BackBufferIndex)->GetResource()->GetID3D12Resource1(),
+		D3D12_RESOURCE_STATE_DEPTH_WRITE,
+		D3D12_RESOURCE_STATE_PIXEL_SHADER_RESOURCE));
+		*/
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/DX12Tasks/PreDepthRenderTask.cpp]: Do you need this? else remove

File: StortSpelProjekt/Engine/src/Renderer/Renderer.cpp [hunk 4]
```diff
@@ -293,6 +294,11 @@ void Renderer::Execute()
 	m_RenderTasks[RENDER_TASK_TYPE::SHADOW]->SetCommandInterfaceIndex(commandInterfaceIndex);
 	m_pThreadPool->AddTask(m_RenderTasks[RENDER_TASK_TYPE::SHADOW], FLAG_THREAD::RENDER);
 
+	// Recording shadowmaps
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/Renderer.cpp]: You probably copy-pasted this from the shadowMap creation. Remove the comment

File: StortSpelProjekt/Engine/src/Renderer/Renderer.cpp [hunk 5]
```diff
@@ -293,6 +294,11 @@ void Renderer::Execute()
...
+	m_RenderTasks[RENDER_TASK_TYPE::PRE_DEPTH]->SetBackBufferIndex(backBufferIndex);
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/Renderer.cpp]: Do you need backBufferIndex?

File: StortSpelProjekt/Engine/src/Renderer/Renderer.cpp [hunk 6]
```diff
@@ -614,6 +621,65 @@ void Renderer::updateMousePicker()
 void Renderer::initRenderTasks()
 {
 	// RenderTasks
+
+#pragma region PreDepthRendering
+
+/* PreDepth rendering without stencil testing */
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/Renderer.cpp]: Tab this comment

File: StortSpelProjekt/Engine/src/Renderer/Renderer.cpp [hunk 7]
```diff
@@ -614,6 +621,65 @@ void Renderer::updateMousePicker()
...
+
+	RenderTask* preDepthRenderTask = new PreDepthRenderTask(
+		m_pDevice5,
+		m_pRootSignature,
+		L"PreDepthVertex.hlsl", L"PreDepthPixel.hlsl",
+		&gpsdPreDepthRenderVector,
+		L"PreDepthRenderingPSO");
+
+	
+	// TODO: remove swapchain, using swapchains render view currently.
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/Renderer.cpp:677]: ?

**Review by jocke1995 — CHANGES_REQUESTED** [review]
Adjust accordingly

File: StortSpelProjekt/Engine/src/Renderer/Renderer.cpp [hunk 8]
```diff
@@ -34,6 +34,7 @@
 #include "MousePicker.h"
 
 // Graphics
+#include "DX12Tasks/PreDepthRenderTask.h"
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/Renderer.cpp]: Change to DepthPrePass

File: StortSpelProjekt/Engine/src/Renderer/Renderer.cpp [hunk 9]
```diff
@@ -614,6 +620,65 @@ void Renderer::updateMousePicker()
 void Renderer::initRenderTasks()
 {
 	// RenderTasks
+
+#pragma region PreDepthRendering
+
+	/* PreDepth rendering without stencil testing */
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/Renderer.cpp]: Depth pre pass Comment

File: StortSpelProjekt/Engine/src/Renderer/Renderer.cpp [hunk 10]
```diff
@@ -614,6 +620,65 @@ void Renderer::updateMousePicker()
...
+	D3D12_GRAPHICS_PIPELINE_STATE_DESC gpsdPreDepthRender = {};
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/Renderer.cpp]: Change name accordingly

File: StortSpelProjekt/Engine/src/Renderer/Renderer.cpp [hunk 11]
```diff
@@ -614,6 +620,65 @@ void Renderer::updateMousePicker()
...
+
+	// DepthStencil
+	preDepthdsd.StencilEnable = false;
+	gpsdPreDepthRender.DepthStencilState = preDepthdsd;
+	gpsdPreDepthRender.DSVFormat = m_pMainDSV->GetDXGIFormat();
+
+	std::vector<D3D12_GRAPHICS_PIPELINE_STATE_DESC*> gpsdPreDepthRenderVector;
+	gpsdPreDepthRenderVector.push_back(&gpsdPreDepthRender);
+
+	RenderTask* preDepthRenderTask = new PreDepthRenderTask(
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/Renderer.cpp]: DepthPreTask

File: StortSpelProjekt/Engine/src/Renderer/Renderer.cpp [hunk 12]
```diff
@@ -1044,6 +1109,7 @@ void Renderer::initRenderTasks()
 	// None atm
 
 	/* ------------------------- DirectQueue Tasks ---------------------- */
+	m_RenderTasks[RENDER_TASK_TYPE::PRE_DEPTH] = preDepthRenderTask;
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/Renderer.cpp]: DEPTH_PREPASS

**Review by jocke1995 — CHANGES_REQUESTED** [review]
Change names

File: StortSpelProjekt/Engine/src/Renderer/DX12Tasks/DX12Task.h [hunk 13]
```diff
@@ -17,6 +17,7 @@ struct ID3D12GraphicsCommandList5;
 // These renderTasks will execute on "all objects"
 enum RENDER_TASK_TYPE
 {
+	DEPTH_PRE_PASS,
```
Comment (jocke1995) [inline StortSpelProjekt/Engine/src/Renderer/DX12Tasks/DX12Task.h:20]: Nice!

**Review by jocke1995 — APPROVED** [review]
Good job

**Review by OtakuFreak13 — APPROVED** [review]
Runs fine and code looks good.

**Pull request merged**